### PR TITLE
Refine CI

### PR DESCRIPTION
**Pull request by Hoverbear**
Use cross platform CI. Linux Stable is red ATM due to a clippy problem, don't block merge for that.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,44 +1,42 @@
...
-  - docker run -d --net=host --name pd --rm pingcap/pd --name "pd" --data-dir "pd" --client-urls "http://127.0.0.1:2379" --advertise-client-urls "http://127.0.0.1:2379"
-  - docker run -d --net=host --name kv --rm pingcap/tikv --pd-endpoints "127.0.0.1:2379" --addr "127.0.0.1:2378" --data-dir "kv"
   - cargo test --all -- --nocapture
-  # Validate benches still work.
-  - cargo bench --all -- --test
+  # We need to switch the daemon of Windows to use Linux docker containers.
+  - if [[ $TRAVIS_OS_NAME == "windows" ]]; then "/c/Program Files/Docker/Docker/DockerCli.exe" -SwitchDaemon .; fi
+  # OS X can't do docker on travis. :(
+  - if [[ $TRAVIS_OS_NAME != "osx" ]]; then docker run -d --net=host --name pd --rm pingcap/pd --name "pd" --data-dir "pd" --client-urls "http://127.0.0.1:2379" --advertise-client-urls "http://127.0.0.1:2379"; fi
+  - if [[ $TRAVIS_OS_NAME != "osx" ]]; then docker run -d --net=host --name kv --rm --ulimit nofile=90000:90000 pingcap/tikv --pd-endpoints "127.0.0.1:2379" --addr "127.0.0.1:2378" --data-dir "kv"; fi
```
Comment (brson) [inline .travis.yml]: These two lines could use the env vars set up earlier for the IPs.

File: .travis.yml [hunk 2]
```diff
@@ -1,44 +1,42 @@
...
-    before_script:
-      - cargo fmt --all -- --check
-      - cargo clippy --all -- -D clippy
-  # Since many users will use nightlies, also test that.
-  - rust: nightly
 
+before_script:
+  - if [[ $TRAVIS_RUST_VERSION == "stable" && $TRAVIS_OS_NAME == "linux" ]]; then cargo fmt -- --check; fi
+  # Currently some crates have lint errors if clippy checks them.
+  - if [[ $TRAVIS_RUST_VERSION == "stable" && $TRAVIS_OS_NAME == "linux" ]]; then cargo clippy -- -D clippy::all; fi
```
Comment (brson) [inline .travis.yml]: Why are these 'before_script' and not part of 'script'?

File: .travis.yml (hunk 1 above)
Comment (Hoverbear) [reply .travis.yml]: I found then Windows breaks. :(

File: .travis.yml (hunk 2 above)
Comment (Hoverbear) [reply .travis.yml]: Ok 👍

**Comment by brson** [comment]
It looks like travis itself is timing out on windows while backing up the cargo cache.

It [_also_ timed out on linux](https://travis-ci.org/tikv/client-rust/jobs/482262844#L1436) but in a way that didn't cause an error.

Maybe for now you could remove the `cache: cargo` line.

**Comment by brson** [comment]
I went ahead and pushed a commit to remove the cargo build cache from travis, see how that affects the windows build.

**Comment by brson** [comment]
Removing the travis build cached fixed the windows errors. I haven't seen linux stable pass yet.

**Comment by Hoverbear** [comment]
This was impacted by https://github.com/pingcap/kvproto/pull/347 which creates several dependency requirements: https://github.com/pingcap/kvproto/issues/354. I'm trying to resolve them.

**Comment by Hoverbear** [comment]
Rebased with signoff because somehow I forgot. 🙈

**Comment by Hoverbear** [comment]
@brson Let me take a look at them...

**Comment by Hoverbear** [comment]
@brson Yes we moved the tests that require a live TiKV to behind a feature flag. Hence the change.

**Pull request merged**